src/bootstrap/builder.rs

```diff
@@ -622,6 +622,10 @@ impl<'a> Builder<'a> {
             }
 
             fn run(self, builder: &Builder<'_>) -> Interned<PathBuf> {
+                if builder.config.download_stage1 {
+                    // Avoid deleting the rustlib/ directory we just copied.
```

Comment:
This comment is pretty confusing to me. There's no copy "just" before this, I guess it's referring to something else, probably a copy in bootstrap.py?

It would also surprise me if the return here is accurate, this is the sysroot libdir, which should be something like sysroot/lib/rustlib/target/lib, not just sysroot?

In any case I'd expect a conditional on the target, compiler host, and stage here? We only download artifacts for build triple after all. (It looks like a helper function of "use downloaded stage 1" may be helpful here, as we have the same problem in several other places it looks like)

Reply:
> There's no copy "just" before this

The order of the code is confusing - the copy happens when `builder.sysroot` is called, in `impl Step for Sysroot`. But there is a copy, without this early return it will copy things to `stage1/` then delete everything in `stage1/lib/rustlib`.

> In any case I'd expect a conditional on the target, compiler host, and stage here?

I have an [assertion for cross-compiling](https://github.com/rust-lang/rust/pull/79540/files#diff-cc65477e7704ede82e44907dcff240ecb0a547eddfadc515bb706e724068f8feR916), I don't want to try supporting that at first. I think we can use a downloaded compiler for any stage, stage1 and stage2 should behave the same.

What's the difference between the target and the compiler host?

> It would also surprise me if the return here is accurate, this is the sysroot libdir, which should be something like sysroot/lib/rustlib/target/lib, not just sysroot?

Good catch, thanks. I don't think it's actually used for anything when stage1 is downloaded, though, maybe just `rustlib/$TARGET/lib/self-contained`.

Reply:
> What's the difference between the target and the compiler host?

The target is essentially the platform std is (cross)compiled for, and the compiler host is the target the compiler is (cross)compiled for.

I think the cross-compiling assertion is slightly wrong; it catches cross-compiling the compiler but not std. I am fine with the initial version not supporting cross-compilation of std, but we'll want to add an assert.

Reply:
> The order of the code is confusing - the copy happens when builder.sysroot is called, in impl Step for Sysroot. But there is a copy, without this early return it will copy things to stage1/ then delete everything in stage1/lib/rustlib.

Ok - that doesn't seem like necessarily just - maybe we can explicitly reference that as you did here in the comment?

Reply:
> maybe we can explicitly reference that as you did here in the comment?

Ok, done.

> I think the cross-compiling assertion is slightly wrong; it catches cross-compiling the compiler but not std. I am fine with the initial version not supporting cross-compilation of std, but we'll want to add an assert.

Std is now skipped altogether when `download-stage1` is set (https://github.com/rust-lang/rust/pull/79540#discussion_r563417852). Ideally this would assert the target is the same as well, but I'm not sure how to do that with an opaque `StepDescription`.